python/ray/_private/event/event_logger.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import pathlib
+import json
+import random
+import string
+import socket
+import os
+import threading
+
+from typing import Dict
+from datetime import datetime
+
+from google.protobuf.json_format import MessageToDict
+
+from ray.core.generated.event_pb2 import Event
+
+
+def get_event_id():
+    return "".join([random.choice(string.hexdigits) for _ in range(36)])
+
+
+class EventLoggerAdapter:
+    def __init__(self, source: Event.SourceType, logger: logging.Logger):
+        """Adapter for the Python logger that's used to emit events.
+
+        This class is threas-safe.
+        """
+        self.logger = logger
+        # Aligned with `event.proto`'s `message Event``
+        self.source = source
+        self.source_hostname = socket.gethostname()
+        self.source_pid = os.getpid()
+
+        # The below fields must be protected by this lock.
+        self.lock = threading.Lock()
+        # {str -> str} typed dict
+        self.global_context = {}
+
+    def set_global_context(self, global_context: Dict[str, str] = None):
+        """Set the global metadata.
+
+        This method overwrites the global metadata if it is called more than once.
+        """
+        with self.lock:
+            self.global_context = {} if not global_context else global_context
+
+    def info(self, message: str, **kwargs):
+        self._emit(Event.Severity.INFO, message, **kwargs)
+
+    def warning(self, message: str, **kwargs):
+        self._emit(Event.Severity.WARNING, message, **kwargs)
+
+    def error(self, message: str, **kwargs):
+        self._emit(Event.Severity.ERROR, message, **kwargs)
+
+    def fatal(self, message: str, **kwargs):
+        self._emit(Event.Severity.FATAL, message, **kwargs)
+
+    def _emit(self, severity: Event.Severity, message: str, **kwargs):
+        # NOTE: Python logger is thread-safe,
+        # so we don't need to protect it using locks.
+        event = Event()
+        event.event_id = get_event_id()
+        event.timestamp = int(datetime.now().timestamp())
+        event.message = message
+        event.severity = severity
+        # TODO(sang): Support event type & schema.
+        event.label = ""
+        event.source_type = self.source
+        event.source_hostname = self.source_hostname
+        event.source_pid = self.source_pid
+        custom_fields = event.custom_fields
+        with self.lock:
```

Comment:
I am a bit confused by the usage of this `global_context`, what would be a use case for this? I couldn't find its usage in the relevant design doc also.

Reply:
Synced offline